go/vt/vttablet/heartbeat/reader.go

```diff
@@ -80,7 +80,6 @@ func (r *Reader) Open(dbc dbconfigs.DBConfigs) {
 	r.mu.Lock()
 	defer r.mu.Unlock()
 	if r.isOpen {
-		log.Fatalf("BUG: Reader object cannot be initialized twice without closing in between: %v", r)
```

Comment:
`actionServeNewType` in `tabletserver.go` is used to handle two cases:
- Going from `StateNotServing` to `StateServing`
- Going from `StateServing` to `StateServing` (i.e. changed type)

The problematic case here is the first one, from`StateNotServing` to `StateServing`. Heartbeat reader/writer is opened and closed based on the tablet type in `serveNewType()`. In the case where the query service shut down due to healthcheck reasons (or otherwise I guess), serving state will go from `StateServing` to `StateNotServing`. 

Previously, we called `Close()` on both heartbeat reader and writer in this case, but since queryservice state is based off of healthchecks, this would cause the tablet to never recover. I changed this in a recent PR, so that the heartbeat reader/writer will continue even when in `StateNotServing`. However, at that point the above problem kicks in -- when we go from `StateNotServing` back to `StateServing`, we try to re-open the proper heartbeat reader/writer based on the tablet type.

This seems like a legitimate workflow. I could add a check for `IsOpen()` in `serveNewType()`, but that seems like overkill.

Does this make sense to you? If so I can document it. If not, we can potentially create a new action type to distinguish between the 2 scenarios I described above.